File: src/lib/pg-schema-compare/protocol.ts
import { z } from "zod";

export const SCHEMA_COMPARE_SCOPE = "postgres16OrdinaryTableProjectionV1";
export const SCHEMA_COMPARE_NORMALIZATION_VERSION = 1;
export const SCHEMA_COMPARE_CHUNK_BYTES = 64 * 1024;
export const SCHEMA_COMPARE_PAGE_BYTES = 1024 * 1024;

const encoder = new TextEncoder();
const bytes = (limit: number) =>
  z
    .string()
    .refine(
      (text) => text.length <= limit && encoder.encode(text).length <= limit,
      "UTF-8 byte limit exceeded",
    );
const identifier = bytes(63).refine(
  (text) => text.length > 0 && !text.includes("\0"),
);
const id = bytes(128).refine((text) => text.length > 0);
const count = z.number().int().nonnegative().max(100_000);

export const schemaCompareEndpoint = z.object({
  connectionId: id,
  schema: identifier,
});
export type SchemaCompareEndpoint = z.infer<typeof schemaCompareEndpoint>;

export const schemaCompareResultIdentity = z.object({
  jobId: id,
  resultId: id,
});
export type SchemaCompareResultIdentity = z.infer<
  typeof schemaCompareResultIdentity
>;

export const schemaCompareNamespace = z.discriminatedUnion("kind", [
  z.object({ kind: z.literal("selected") }),
  z.object({ kind: z.literal("external"), schema: identifier }),
]);
export const schemaCompareQualifiedName = z.object({
  namespace: schemaCompareNamespace,
  name: identifier,
});
export type SchemaCompareQualifiedName = z.infer<
  typeof schemaCompareQualifiedName
>;

export const schemaCompareRelationIdentity = z.object({
  kind: z.enum([
    "table",
    "partitionedTable",
    "foreignTable",
    "view",
    "materializedView",
    "sequence",
    "composite",
  ]),
  name: identifier,
});

export const schemaCompareFieldPath = z.discriminatedUnion("kind", [
  z.object({
    kind: z.literal("table"),
    field: z.enum(["persistence", "comment"]),
  }),
  z.object({
    kind: z.literal("column"),
    name: identifier,
    field: z.enum([
      "position",
      "type",
      "typeModifier",
      "arrayDimensions",
      "nullable",
      "default",
      "generatedKind",
      "generatedExpression",
      "identity",
      "collation",
      "comment",
    ]),
  }),
  z.object({
    kind: z.literal("constraint"),
    name: identifier,
    field: z.enum([
      "kind",
      "keys",
      "referencedTable",
      "referencedKeys",
      "updateAction",
      "deleteAction",
      "deleteColumns",
      "matchMode",
      "deferrable",
      "initiallyDeferred",
      "validated",
      "noInherit",
      "expression",
      "equalityOperators",
      "exclusionOperators",
    ]),
  }),
  z.object({
    kind: z.literal("index"),
    name: identifier,
    owner: identifier.nullable(),
    field: z.enum([
      "accessMethod",
      "unique",
      "nullsNotDistinct",
      "immediate",
      "keyCount",
      "includedColumns",
      "predicate",
      "relationOptions",
      "valid",
      "ready",
      "live",
    ]),
  }),
  z.object({
    kind: z.literal("indexKey"),
    name: identifier,
    owner: identifier.nullable(),
    position: z.number().int().nonnegative().max(65_535),
    field: z.enum([
      "kind",
      "column",
      "expression",
      "sortOptions",
      "opclass",
      "opclassOptions",
      "collation",
    ]),
  }),
]);
export type SchemaCompareFieldPath = z.infer<typeof schemaCompareFieldPath>;

export const schemaCompareValueRef = z.object({
  side: z.enum(["source", "target"]),
  valueId: z.number().int().nonnegative().max(99_999),
  rawBytes: z
    .number()
    .int()
    .nonnegative()
    .max(256 * 1024),
  valueKind: z.enum([
    "null",
    "text",
    "boolean",
    "integer",
    "qualifiedName",
    "orderedNames",
    "orderedReferences",
    "operatorSignatures",
  ]),
});
export type SchemaCompareValueRef = z.infer<typeof schemaCompareValueRef>;
export const schemaCompareValueRequest = z.object({
  identity: schemaCompareResultIdentity,
  value: schemaCompareValueRef,
  offset: z
    .number()
    .int()
    .nonnegative()
    .max(256 * 1024),
});
export type SchemaCompareValueRequest = z.infer<
  typeof schemaCompareValueRequest
>;

export const schemaCompareChunk = z
  .object({
    responseId: id,
    identity: schemaCompareResultIdentity,
    value: schemaCompareValueRef,
    offset: z
      .number()
      .int()
      .nonnegative()
      .max(256 * 1024),
    text: bytes(SCHEMA_COMPARE_CHUNK_BYTES),
    nextOffset: z
      .number()
      .int()
      .nonnegative()
      .max(256 * 1024),
    complete: z.boolean(),
  })
  .refine(
    (chunk) =>
      chunk.text.length <= SCHEMA_COMPARE_CHUNK_BYTES &&
      chunk.nextOffset === chunk.offset + encoder.encode(chunk.text).length &&
      chunk.nextOffset <= chunk.value.rawBytes &&
      chunk.complete === (chunk.nextOffset === chunk.value.rawBytes),
    "Invalid chunk offsets",
  );
export type SchemaCompareChunk = z.infer<typeof schemaCompareChunk>;

const incomparableReason = z.enum([
  "expressionOutsideSubset",
  "renderingVersionDifference",
  "externalDependency",
  "unknownAccessMethod",
  "excludedCounterpart",
  "excludedObject",
]);
const sourceValueRef = schemaCompareValueRef.extend({
  side: z.literal("source"),
});
const targetValueRef = schemaCompareValueRef.extend({
  side: z.literal("target"),
});
const fieldSummary = { path: schemaCompareFieldPath };
export const schemaCompareFieldSummary = z.union([
  z.strictObject({
    ...fieldSummary,
    kind: z.literal("equal"),
    source: sourceValueRef,
    target: targetValueRef,
  }),
  z.strictObject({
    ...fieldSummary,
    kind: z.literal("changed"),
    source: sourceValueRef,
    target: targetValueRef,
  }),
  z.strictObject({
    ...fieldSummary,
    kind: z.literal("sourceOnly"),
    source: sourceValueRef,
  }),
  z.strictObject({
    ...fieldSummary,
    kind: z.literal("targetOnly"),
    target: targetValueRef,
  }),
  z.strictObject({
    ...fieldSummary,
    kind: z.literal("notComparable"),
    reason: incomparableReason,
    source: sourceValueRef,
    target: targetValueRef,
  }),
  z.strictObject({
    ...fieldSummary,
    kind: z.literal("notComparable"),
    reason: incomparableReason,
    source: sourceValueRef,
  }),
  z.strictObject({
    ...fieldSummary,
    kind: z.literal("notComparable"),
    reason: incomparableReason,
    target: targetValueRef,
  }),
]);
export type SchemaCompareFieldSummary = z.infer<
  typeof schemaCompareFieldSummary
>;
const objectCounts = {
  fieldCount: count,
  changedFields: count,
  incomparableFields: count,
};
export const schemaCompareObjectSummary = z.union([
  z.strictObject({
    ...objectCounts,
    kind: z.literal("equal"),
    source: schemaCompareRelationIdentity,
    target: schemaCompareRelationIdentity,
  }),
  z.strictObject({
    ...objectCounts,
    kind: z.literal("changed"),
    source: schemaCompareRelationIdentity,
    target: schemaCompareRelationIdentity,
  }),
  z.strictObject({
    ...objectCounts,
    kind: z.literal("sourceOnly"),
    source: schemaCompareRelationIdentity,
  }),
  z.strictObject({
    ...objectCounts,
    kind: z.literal("targetOnly"),
    target: schemaCompareRelationIdentity,
  }),
  z.strictObject({
    ...objectCounts,
    kind: z.literal("notComparable"),
    reason: incomparableReason,
    source: schemaCompareRelationIdentity,
    target: schemaCompareRelationIdentity,
  }),
  z.strictObject({
    ...objectCounts,
    kind: z.literal("notComparable"),
    reason: incomparableReason,
    source: schemaCompareRelationIdentity,
  }),
  z.strictObject({
    ...objectCounts,
    kind: z.literal("notComparable"),
    reason: incomparableReason,
    target: schemaCompareRelationIdentity,
  }),
]);
const pageMetadata = {
  responseId: id,
  identity: schemaCompareResultIdentity,
  offset: count,
  nextOffset: count.nullable(),
};
export const schemaCompareObjectPage = z.object({
  ...pageMetadata,
  items: z.array(schemaCompareObjectSummary).max(100),
});
export const schemaCompareFieldPage = z.object({
  ...pageMetadata,
  items: z.array(schemaCompareFieldSummary).max(100),
});

export const schemaCompareError = z.discriminatedUnion("kind", [
  z.object({ kind: z.literal("busy") }),
  z.object({
    kind: z.literal("limitExceeded"),
    limit: z.enum([
      "inventory",
      "tables",
      "childFacts",
      "fieldBytes",
      "endpointBytes",
      "resultBytes",
      "pageBytes",
      "pageItems",
      "allocation",
    ]),
  }),
  z.object({
    kind: z.literal("unsupportedVersion"),
    side: z.enum(["source", "target"]),
    version: bytes(128),
  }),
  z.object({
    kind: z.literal("unsupportedEngine"),
    side: z.enum(["source", "target"]),
  }),
  z.object({ kind: z.literal("unavailable") }),
  z.object({ kind: z.literal("invalidRequest") }),
  z.object({ kind: z.literal("captureChanged") }),
  z.object({ kind: z.literal("cancelled") }),
  z.object({ kind: z.literal("deadlineExceeded") }),
]);
export type SchemaCompareError = z.infer<typeof schemaCompareError>;

const statusMetadata = {
  jobId: id,
  requestId: id,
  source: schemaCompareEndpoint,
  target: schemaCompareEndpoint,
  sourceObjects: count,
  targetObjects: count,
};
export const schemaCompareStatus = z.discriminatedUnion("phase", [
  z.strictObject({ ...statusMetadata, phase: z.literal("resolving") }),
  z.strictObject({ ...statusMetadata, phase: z.literal("readingSource") }),
  z.strictObject({ ...statusMetadata, phase: z.literal("readingTarget") }),
  z.strictObject({ ...statusMetadata, phase: z.literal("readingBoth") }),
  z.strictObject({ ...statusMetadata, phase: z.literal("comparing") }),
  z.strictObject({
    ...statusMetadata,
    phase: z.literal("completed"),
    resultId: id,
  }),
  z.strictObject({ ...statusMetadata, phase: z.literal("cancelling") }),
  z.strictObject({ ...statusMetadata, phase: z.literal("cancelled") }),
  z.strictObject({
    ...statusMetadata,
    phase: z.literal("failed"),
    failure: schemaCompareError,
  }),
]);
export type SchemaCompareStatus = z.infer<typeof schemaCompareStatus>;

export const schemaCompareCaptureMetadata = z.object({
  endpoint: schemaCompareEndpoint,
  serverVersion: bytes(256),
  serverVersionNum: z.number().int().nonnegative().max(4_294_967_295),
  capturedAt: bytes(64),
});
const schemaCompareExcludedCategory = z.enum([
  "otherRelations",
  "routines",
  "sequences",
  "typesAndDomains",
  "policies",
  "grants",
  "triggers",
  "rules",
  "extensions",
  "databaseObjects",
  "identitySequenceConfiguration",
  "storageSecurityOwnershipReplication",
  "indexPlacementClusteringReplicaIdentity",
]);
export const schemaCompareEligibility = z.discriminatedUnion("kind", [
  z.object({ kind: z.literal("eligible") }),
  z.object({
    kind: z.literal("excluded"),
    reason: z.enum([
      "partitioned",
      "inherited",
      "foreign",
      "extensionOwned",
      "otherKind",
    ]),
  }),
]);
export const schemaCompareCoverage = z.object({
  scope: z.literal(SCHEMA_COMPARE_SCOPE),
  normalizationVersion: z.literal(SCHEMA_COMPARE_NORMALIZATION_VERSION),
  excludedRelations: count,
  incomparableFields: count,
  excludedCategories: z.array(schemaCompareExcludedCategory).max(13),
});
export const schemaCompareInventoryEntry = z.object({
  identity: schemaCompareRelationIdentity,
  eligibility: schemaCompareEligibility,
});
export const schemaCompareOperatorSignature = z.object({
  operator: schemaCompareQualifiedName,
  leftType: schemaCompareQualifiedName.nullable(),
  rightType: schemaCompareQualifiedName.nullable(),
});
export type SchemaCompareCoverage = z.infer<typeof schemaCompareCoverage>;
export type SchemaCompareCaptureMetadata = z.infer<
  typeof schemaCompareCaptureMetadata
>;
export type SchemaCompareInventoryEntry = z.infer<
  typeof schemaCompareInventoryEntry
>;
export type SchemaCompareOperatorSignature = z.infer<
  typeof schemaCompareOperatorSignature
>;

export const schemaCompareMetadata = z.object({
  identity: schemaCompareResultIdentity,
  source: schemaCompareCaptureMetadata,
  target: schemaCompareCaptureMetadata,
  consistency: z.enum(["sharedTransaction", "independentTransactions"]),
  coverage: schemaCompareCoverage,
});
export type SchemaCompareMetadata = z.infer<typeof schemaCompareMetadata>;

export const schemaCompareStartRequest = z.object({
  requestId: id,
  source: schemaCompareEndpoint,
  target: schemaCompareEndpoint,
});
export type SchemaCompareStartRequest = z.infer<
  typeof schemaCompareStartRequest
>;
export const schemaCompareResultRequest = z.object({
  identity: schemaCompareResultIdentity,
  source: schemaCompareEndpoint,
  target: schemaCompareEndpoint,
});
export type SchemaCompareResultRequest = z.infer<
  typeof schemaCompareResultRequest
>;
export const schemaCompareReadRequest = z.discriminatedUnion("kind", [
  z.object({ kind: z.literal("metadata") }),
  z.object({ kind: z.literal("objects"), offset: count }),
  z.object({
    kind: z.literal("fields"),
    object: schemaCompareRelationIdentity,
    offset: count,
  }),
  z.object({
    kind: z.literal("eligibility"),
    object: schemaCompareRelationIdentity,
    side: z.enum(["source", "target"]),
  }),
  z.object({
    kind: z.literal("value"),
    value: schemaCompareValueRef,
    offset: z
      .number()
      .int()
      .nonnegative()
      .max(256 * 1024),
  }),
]);
export type SchemaCompareReadRequest = z.infer<typeof schemaCompareReadRequest>;
const excludedCount = z.object({
  category: schemaCompareExcludedCategory,
  count,
  complete: z.boolean(),
});
export const schemaCompareMetadataPage = z.object({
  responseId: id,
  identity: schemaCompareResultIdentity,
  detail: z.object({
    metadata: schemaCompareMetadata,
    kind: z.enum([
      "equal",
      "changed",
      "sourceOnly",
      "targetOnly",
      "notComparable",
    ]),
    objectCount: count,
    sourceExcludedCounts: z.array(excludedCount).max(13),
    targetExcludedCounts: z.array(excludedCount).max(13),
  }),
});
export const schemaCompareEligibilityPage = z.object({
  responseId: id,
  identity: schemaCompareResultIdentity,
  detail: z.object({
    object: schemaCompareRelationIdentity,
    side: z.enum(["source", "target"]),
    eligibility: schemaCompareEligibility,
  }),
});

File: src-tauri/src/postgres/schema_compare/protocol.rs
use serde::{Deserialize, Serialize};

pub const NORMALIZATION_VERSION: u32 = 1;
pub const SCOPE: &str = "postgres16OrdinaryTableProjectionV1";

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Endpoint {
    pub connection_id: String,
    pub schema: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResultIdentity {
    pub job_id: String,
    pub result_id: String,
}

impl ResultIdentity {
    pub fn validate(&self) -> Result<(), CompareError> {
        if [&self.job_id, &self.result_id]
            .iter()
            .any(|s| s.is_empty() || s.len() > 128)
        {
            return Err(CompareError::InvalidRequest);
        }
        Ok(())
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum Side {
    Source,
    Target,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum RelationKind {
    Table,
    PartitionedTable,
    ForeignTable,
    View,
    MaterializedView,
    Sequence,
    Composite,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum Exclusion {
    Partitioned,
    Inherited,
    Foreign,
    ExtensionOwned,
    OtherKind,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum Eligibility {
    Eligible,
    Excluded { reason: Exclusion },
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RelationIdentity {
    pub kind: RelationKind,
    pub name: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InventoryEntry {
    pub identity: RelationIdentity,
    pub eligibility: Eligibility,
}

/// Selected schemas map structurally to one namespace. External references
/// keep their exact schema. Raw SQL text is never passed through this mapping.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum Namespace {
    Selected,
    External { schema: String },
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct QualifiedName {
    pub namespace: Namespace,
    pub name: String,
}

impl QualifiedName {
    pub fn in_endpoint(endpoint: &Endpoint, schema: &str, name: &str) -> Self {
        Self {
            namespace: if schema == endpoint.schema {
                Namespace::Selected
            } else {
                Namespace::External {
                    schema: schema.into(),
                }
            },
            name: name.into(),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum TableField {
    Persistence,
    Comment,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum ColumnField {
    Position,
    Type,
    TypeModifier,
    ArrayDimensions,
    Nullable,
    Default,
    GeneratedKind,
    GeneratedExpression,
    Identity,
    Collation,
    Comment,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum ConstraintField {
    Kind,
    Keys,
    ReferencedTable,
    ReferencedKeys,
    UpdateAction,
    DeleteAction,
    DeleteColumns,
    MatchMode,
    Deferrable,
    InitiallyDeferred,
    Validated,
    NoInherit,
    Expression,
    EqualityOperators,
    ExclusionOperators,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum IndexField {
    AccessMethod,
    Unique,
    NullsNotDistinct,
    Immediate,
    KeyCount,
    IncludedColumns,
    Predicate,
    RelationOptions,
    Valid,
    Ready,
    Live,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum IndexKeyField {
    Kind,
    Column,
    Expression,
    SortOptions,
    Opclass,
    OpclassOptions,
    Collation,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum FieldPath {
    Table {
        field: TableField,
    },
    Column {
        name: String,
        field: ColumnField,
    },
    Constraint {
        name: String,
        field: ConstraintField,
    },
    Index {
        name: String,
        owner: Option<String>,
        field: IndexField,
    },
    IndexKey {
        name: String,
        owner: Option<String>,
        position: u16,
        field: IndexKeyField,
    },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum IncomparableReason {
    ExpressionOutsideSubset,
    RenderingVersionDifference,
    ExternalDependency,
    UnknownAccessMethod,
    ExcludedCounterpart,
    ExcludedObject,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OperatorSignature {
    pub operator: QualifiedName,
    pub left_type: Option<QualifiedName>,
    pub right_type: Option<QualifiedName>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum CompareError {
    Busy,
    LimitExceeded { limit: Limit },
    UnsupportedVersion { side: Side, version: String },
    UnsupportedEngine { side: Side },
    Unavailable,
    InvalidRequest,
    CaptureChanged,
    Cancelled,
    DeadlineExceeded,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum Limit {
    Inventory,
    Tables,
    ChildFacts,
    FieldBytes,
    EndpointBytes,
    ResultBytes,
    PageBytes,
    PageItems,
    Allocation,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "phase", rename_all = "camelCase")]
pub enum StatusState {
    Resolving,
    ReadingSource,
    ReadingTarget,
    ReadingBoth,
    Comparing,
    Completed {
        #[serde(rename = "resultId")]
        result_id: String,
    },
    Cancelling,
    Cancelled,
    Failed {
        failure: CompareError,
    },
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CaptureMetadata {
    pub endpoint: Endpoint,
    pub server_version: String,
    pub server_version_num: u32,
    pub captured_at: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum SnapshotConsistency {
    SharedTransaction,
    IndependentTransactions,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ComparisonMetadata {
    pub identity: ResultIdentity,
    pub source: CaptureMetadata,
    pub target: CaptureMetadata,
    pub consistency: SnapshotConsistency,
    pub coverage: Coverage,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Coverage {
    pub scope: String,
    pub normalization_version: u32,
    pub excluded_relations: u32,
    pub incomparable_fields: u32,
    pub excluded_categories: Vec<ExcludedCategory>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum ExcludedCategory {
    OtherRelations,
    Routines,
    Sequences,
    TypesAndDomains,
    Policies,
    Grants,
    Triggers,
    Rules,
    Extensions,
    DatabaseObjects,
    IdentitySequenceConfiguration,
    StorageSecurityOwnershipReplication,
    IndexPlacementClusteringReplicaIdentity,
}

/// List/status payloads contain counts and endpoint metadata, never values.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Status {
    pub job_id: String,
    pub request_id: String,
    pub source: Endpoint,
    pub target: Endpoint,
    pub source_objects: u32,
    pub target_objects: u32,
    #[serde(flatten)]
    pub state: StatusState,
}

#[cfg(test)]
mod tests {
    use super::*;

    fn status(state: StatusState) -> Status {
        Status {
            job_id: "job".into(),
            request_id: "request".into(),
            source: Endpoint {
                connection_id: "source".into(),
                schema: "public".into(),
            },
            target: Endpoint {
                connection_id: "target".into(),
                schema: "public".into(),
            },
            source_objects: 1,
            target_objects: 1,
            state,
        }
    }

    #[test]
    fn status_serialization_binds_terminal_payloads_to_their_phase() {
        let completed = serde_json::to_value(status(StatusState::Completed {
            result_id: "result".into(),
        }))
        .unwrap();
        assert_eq!(completed["phase"], "completed");
        assert_eq!(completed["resultId"], "result");
        assert!(completed.get("failure").is_none());

        let failed = serde_json::to_value(status(StatusState::Failed {
            failure: CompareError::CaptureChanged,
        }))
        .unwrap();
        assert_eq!(failed["phase"], "failed");
        assert_eq!(failed["failure"]["kind"], "captureChanged");
        assert!(failed.get("resultId").is_none());
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct StartRequest {
    pub request_id: String,
    pub source: Endpoint,
    pub target: Endpoint,
}
impl StartRequest {
    /// Timestamped caller IDs make an evicted/expired start fail closed. A
    /// retry may reconcile retained history, but cannot create work again after
    /// the ten-minute request ledger has forgotten it.
    pub(crate) fn fresh(&self) -> bool {
        let Some((millis, nonce)) = self.request_id.split_once(':') else {
            return false;
        };
        let Ok(millis) = millis.parse::<i64>() else {
            return false;
        };
        let age = chrono::Utc::now().timestamp_millis().saturating_sub(millis);
        !nonce.is_empty() && (-5_000..60_000).contains(&age)
    }
    pub(crate) fn validate(&self) -> Result<(), CompareError> {
        if self.request_id.is_empty()
            || self.request_id.len() > 128
            || [&self.source, &self.target].iter().any(|e| {
                e.connection_id.is_empty()
                    || e.connection_id.len() > 128
                    || e.schema.is_empty()
                    || e.schema.len() > 63
                    || e.schema.contains('\0')
            })
        {
            return Err(CompareError::InvalidRequest);
        }
        Ok(())
    }
}

/// Every result read binds the immutable result to both exact schema endpoints.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ResultRequest {
    pub identity: ResultIdentity,
    pub source: Endpoint,
    pub target: Endpoint,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum ReadRequest {
    Metadata,
    Objects {
        offset: u32,
    },
    Fields {
        object: RelationIdentity,
        offset: u32,
    },
    Eligibility {
        object: RelationIdentity,
        side: Side,
    },
    Value {
        value: super::values::ValueRef,
        offset: u32,
    },
}
